common/tor.c

```diff
@@ -58,175 +54,198 @@ struct tor_service_reaching {
 	size_t hlen;
 };
 
-
-static struct io_plan *io_tor_connect_create_onion_finished(struct io_conn *conn, struct tor_service_reaching *reach)
+static struct io_plan *io_tor_connect_create_onion_finished(struct io_conn
+							    *conn,
+							    struct
+							    tor_service_reaching
+							    *reach)
 {
-    u8 tor_dec_bytes[TOR_V3_ADDRLEN];
-
-	if(reach->hlen ==  MAX_TOR_ONION_V2_ADDR_LEN ) {
-	size_t n = tal_count(reach->ld->wireaddrs);
-	tal_resize(&reach->ld->wireaddrs, n+1);
-	reach->ld->wireaddrs[n].type = ADDR_TYPE_TOR_V2;
-	reach->ld->wireaddrs[n].addrlen = TOR_V2_ADDRLEN;
-	reach->ld->wireaddrs[n].port = reach->ld->portnum;
-	b32_decode((u8 *)tor_dec_bytes,(u8 *)reach->buffer,2);
-	memcpy(&reach->ld->wireaddrs[n].addr,tor_dec_bytes, reach->ld->wireaddrs[n].addrlen);
-
+	char *temp_char;
+	const tal_t *tmpctx = tal_tmpctx(NULL);
+	if (reach->hlen == MAX_TOR_ONION_V2_ADDR_LEN) {
+		size_t n = tal_count(reach->ld->wireaddrs);
+		tal_resize(&reach->ld->wireaddrs, n + 1);
+		temp_char = tal_fmt(tmpctx, "%.56s.onion", reach->buffer);
+		parse_wireaddr(temp_char, &reach->ld->wireaddrs[n],
+			       reach->ld->portnum, NULL);
 	}
-	/*on the other hand we can stay connected until ln finish to keep onion alive and then vanish*/
+	/*on the other hand we can stay connected until ln finish to keep onion alive and then vanish */
 	//because when we run with Detach flag as we now do every start of LN creates a new addr while the old
 	//stays valid until reboot this might not be desired so we can also drop Detach and use the
 	//read_partial to keep it open until LN drops
 	//FIXME: SAIBATO
 	//return io_read_partial(conn, reach->p, 1 ,&reach->hlen, io_tor_connect_create_onion_finished, reach);
+	tal_free(tmpctx);
 	return io_close(conn);
 }
 
-
-static struct io_plan *io_tor_connect_after_create_onion(struct io_conn *conn, struct tor_service_reaching *reach)
+static struct io_plan *io_tor_connect_after_create_onion(struct io_conn *conn,
+							 struct
+							 tor_service_reaching
+							 *reach)
 {
 
-	reach->p = reach->p+reach->hlen;
-
- if (!strstr((char *)reach->buffer, "ServiceID=")) {
-	if (reach->hlen == 0)
-		return io_close(conn);
-	return io_read_partial(conn, reach->p, 1 ,&reach->hlen, io_tor_connect_after_create_onion, reach);
-	}
-	else
-	{
-	memset(reach->buffer,0,sizeof(reach->buffer));
-	return io_read_partial(conn, reach->buffer, MAX_TOR_ONION_V2_ADDR_LEN ,&reach->hlen, io_tor_connect_create_onion_finished, reach);
+	reach->p = reach->p + reach->hlen;
+
+	if (!strstr((char *)reach->buffer, "ServiceID=")) {
```

Comment:
Note that the previous function which triggers this function is `io_tor_connect_make_onion`, which loads some data into the buffer.  But that function loads `reach->hlen` with the size of the data it will *write*.  This function then adds `reach->p = reach->p + reach->hlen`, effectively putting the read buffer after the written data.  Is that OK?

Reply:
was the only way got the read_partial not stuck ;-) maybe i  find a better way